### PR TITLE
make THPPointer have explicit constructors

### PR DESCRIPTION
**Pull request by killeent**
Previously, THPPointer did not have its constructors marked as explicit. As a result, something like:

```
THTensor *someIntermediateResult = THTensor_(new)(LIBRARY_STATE_NOARGS);
// do some things to someIntermediateResult
THPTensor wrapped = someIntermediateResult; // calls implicit constructor
```

The fact that `THPTensor` takes "ownership" of the `THTensor` in this case is obfuscated, which could lead to double frees. 

This diff marks all the `THPTensor` constructors explicit, preventing this type of implicit constructor from a pointer. I followed it with the following codemod:

```
codemod -m -d torch/ --extensions h,hpp,cpp,cwrap \
    '(TH[a-zA-Z_]+Ptr [a-zA-Z_]+) = ([^;]*);'  \
    '\1(\2);'
```

To replace all assignment calls with calls to the explicit constructor. There were a couple of other places where I need to fix this, e.g. in the `THPPlugin.py` allocation template. 

Test Plan: Run Tests locally

### CODE REVIEW

File: torch/csrc/utils.cpp [hunk 1]
```diff
@@ -21,31 +21,31 @@ int THPUtils_getCallable(PyObject *arg, PyObject **result) {
   return 1;
 }
 
-THLongStoragePtr THPUtils_unpackSize(PyObject *arg) {
+THLongStorage* THPUtils_unpackSize(PyObject *arg) {
```
Comment (colesbury) [inline torch/csrc/utils.cpp]: Why did you change this?

File: torch/csrc/utils/object_ptr.h [hunk 2]
```diff
@@ -5,9 +5,9 @@
 template<class T>
 class THPPointer {
 public:
-  THPPointer(): ptr(nullptr) {};
-  THPPointer(T *ptr): ptr(ptr) {};
-  THPPointer(THPPointer &&p) { free(); ptr = p.ptr; p.ptr = nullptr; };
+  explicit THPPointer(): ptr(nullptr) {};
```
Comment (colesbury) [inline torch/csrc/utils/object_ptr.h]: I suspect we should probably follow the convention of std::unique_ptr (http://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr) and only have line 9 be marked explicit.

**Comment by killeent** [comment]
@colesbury I think it was due to the fact that the move constructor had been marked explicit. By removing the keyword, I no longer need to pass the Tensor as a raw pointer return type in order for it to compile.

**Review by soumith — APPROVED** [review]
